src/language/__tests__/parser-test.ts

```diff
@@ -224,6 +228,463 @@ describe('Parser', () => {
+                    loc: { start: 8, end: 9 },
+                    element: undefined,
+                  },
+                },
+                selectionSet: undefined,
+              },
+            ],
+          },
+        },
+      ],
+    });
+  });
+
+  it('parses field with optional list elements', () => {
+    const document = '{ field[?] }';
+    const result = parse(document, {
+      experimentalClientControlledNullability: true,
+    });
+
+    expectJSON(result).toDeepEqual({
+      kind: Kind.DOCUMENT,
+      loc: { start: 0, end: 12 },
+      definitions: [
+        {
+          kind: Kind.OPERATION_DEFINITION,
+          loc: { start: 0, end: 12 },
+          operation: 'query',
+          name: undefined,
+          variableDefinitions: [],
+          directives: [],
+          selectionSet: {
+            kind: Kind.SELECTION_SET,
+            loc: { start: 0, end: 12 },
+            selections: [
+              {
+                kind: Kind.FIELD,
+                loc: { start: 2, end: 10 },
+                alias: undefined,
+                name: {
+                  kind: Kind.NAME,
+                  loc: { start: 2, end: 7 },
+                  value: 'field',
+                },
+                arguments: [],
+                directives: [],
+                required: {
+                  kind: Kind.LIST_NULLABILITY,
+                  loc: { start: 7, end: 10 },
+                  element: {
+                    kind: Kind.OPTIONAL_DESIGNATOR,
+                    loc: { start: 8, end: 9 },
+                    element: undefined,
+                  },
+                },
+                selectionSet: undefined,
+              },
+            ],
+          },
+        },
+      ],
+    });
+  });
+
+  it('parses field with required list', () => {
+    const document = '{ field[]! }';
+    const result = parse(document, {
+      experimentalClientControlledNullability: true,
+    });
+
+    expectJSON(result).toDeepEqual({
+      kind: Kind.DOCUMENT,
+      loc: { start: 0, end: 12 },
+      definitions: [
+        {
+          kind: Kind.OPERATION_DEFINITION,
+          loc: { start: 0, end: 12 },
+          operation: 'query',
+          name: undefined,
+          variableDefinitions: [],
+          directives: [],
+          selectionSet: {
+            kind: Kind.SELECTION_SET,
+            loc: { start: 0, end: 12 },
+            selections: [
+              {
+                kind: Kind.FIELD,
+                loc: { start: 2, end: 10 },
+                alias: undefined,
+                name: {
+                  kind: Kind.NAME,
+                  loc: { start: 2, end: 7 },
+                  value: 'field',
+                },
+                arguments: [],
+                directives: [],
+                selectionSet: undefined,
+                required: {
+                  kind: Kind.REQUIRED_DESIGNATOR,
+                  loc: { start: 9, end: 10 },
```

Comment:
Invalid, should be
```suggestion
                  loc: { start: 7, end: 10 },
```
An outer node should have a wider range and include all its inner nodes.

Reply:
I working on a review commit that will fix that.